frontend/model/captureLogs.js

```diff
@@ -0,0 +1,250 @@
+
+function captureLog (type, ...msg) {
+  const logEntry = `${Date.now()}_${Math.floor(Math.random() * 100000)}` // uuid
+
+  // detect when is an Error and capture it properly
+  // ex: uncaught Vue errors or custom try/catch errors.
+  msg = msg.map((m) => m instanceof Error ? (m.stack || m.message) : m)
+
+  giLSset(logEntry, JSON.stringify({
+    type,
+    msg,
+    prev: lastEntry,
+    timestamp: new Date().toISOString()
+  }))
+
+  lastEntry = logEntry
+  entriesCount++
+
+  giLSset('lastEntry', logEntry)
+  giLSset('count', entriesCount)
+
+  verifyEntryIsMarker()
+
+  // To avoid infinite loop because we log all selector calls, we run sbp calls
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyEntryIsMarker () {
+  if (entriesCount % ENTRIES_MARKER_NTH === 0) {
+    const markers = getMarkers()
+    // Save entry as a marker to be later deleted when logs are too big.
+    markers.push(lastEntry)
+    giLSset('markers', JSON.stringify(markers))
+  }
+}
+
+function verifyLogsSize () {
+  let lastEntriesCount = null
+  let failedDeletionCount = 0
+  let markers = null
+
+  // Delete nth oldest logs recursively. This scenario can happen when the
+  // entries limit is changed. Example: There are 400 logs and the limit is 500.
+  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+  // We do it recursively in chunks of 25 until there's only 75 logs again.
+  while (entriesCount >= ENTRIES_LIMIT && failedDeletionCount < 3) {
+    markers = markers || getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+
+    if (entriesCount === lastEntriesCount) { failedDeletionCount++ }
+    lastEntriesCount = entriesCount
+
+    do {
+      const log = JSON.parse(giLSget(oldestEntry)) || {}
+      giLSremove(oldestEntry)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+  }
+
+  if (failedDeletionCount < 3) {
+    giLSset('markers', JSON.stringify(markers))
+    giLSset('count', entriesCount)
+  } else {
+    // - There are too many entriesCount, however, for some unknown reason
+    //   we weren't able to remove them.
+    // - Capture a new log error manually (withou console.error) so
+    //   that verifyLogsSize() doesn't get called again. There's no need
+    //   for that and it would cause a loop of failed deletions.
+    captureLog('error', 'verifyLogsSize() failed! Unable to delete oldest logs.')
+
+    // To avoid infinite loops or memory leaks, let's just pause the capturing.
+    captureLogsPause()
+    // Or... if we want to be extreme, clear the logs and have a fresh start?
+    // clearLogs()
+  }
+}
```

Comment:
I've made a few slight changes to this function to make it compatible with the modified version of `verifyLogsConfigs()` and also some small optimizations:

```js
function verifyLogsSize () {
  let lastEntriesCount = null
  // Delete nth oldest logs recursively. This scenario can happen when the
  // entries limit is changed. Example: There are 400 logs and the limit is 500.
  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
  // We do it recursively in chunks of 25 until there's only 75 logs again.
  while (entriesCount > ENTRIES_LIMIT) {
    if (entriesCount === lastEntriesCount) {
      // There are too many entriesCount, however, for some unknown reason
      // we weren't able to remove them. Call clearLogs() to reset everything.
      clearLogs()
      // Capture a new log error manually (withou console.error) so
      // that verifyLogsSize() doesn't get called again. There's no need
      // for that and it would cause a loop of failed deletions.
      captureLog('error', 'verifyLogsSize(): unable to delete oldest logs, had to clear them!')
      return // return immediately
    }
    lastEntriesCount = entriesCount

    // since this function is run on every single log, we getMarkers()
    // inside the loop so that most times we aren't touching localStorage
    let markers = getMarkers()
    let toDelete = ENTRIES_MARKER_NTH
    let oldestEntry = markers.shift() // the oldest marker

    do {
      const log = JSON.parse(giLSget(oldestEntry)) || {}
      giLSremove(oldestEntry)
      entriesCount--
      toDelete--
      oldestEntry = log.prev
    } while (toDelete && oldestEntry)
    // set inside the while loop so that most times we aren't touching localStorage
    giLSset('markers', JSON.stringify(markers))
    giLSset('count', entriesCount)
  }
}
```

Reply:
Code changed! The only difference from your suggestion is:

1. Using console.error. It's now safe to use `console.error` because the logs were reset so we won't have problems with the limit.

```diff
- captureLog('error', 'verifyLogsSize(): unable to delete oldest logs, had to clear them!')
+ console.error('verifyLogsSize(): un...')
```

2.  The usage of `getMarkers()`.

```diff
// since this function is run on every single log, we getMarkers()
// inside the loop so that most times we aren't touching localStorage
- let markers = getMarkers()
+ markers = markers || getMarkers()
```

I think you are contradicting yourself. This function is not run on every single log. It's only run when the `entryCount` reached the limit. And because this is a loop, that we only need to access the markers once. That's why I did `markers = markers || getMarkers()`. This means, if there are markers, use them, otherwise, let's get them from LS. That way, in this loop we only hit the localStorage once: the first time we enter this while loop.

Reply:
OK this is fine but I think the version of `verifyLogsConfigs()` that I wrote should be incorporated as well (you can remove the part that calculates the `entriesCount` from scratch.

Reply:
I did incorporate it and improved it even more. Maybe you were reviewing old changes?